src/protocol.rs

```diff
@@ -339,9 +341,50 @@ impl ParticipantIdentifier {
     }
 }
 
-impl ProofContext for Vec<ParticipantIdentifier> {
-    fn as_bytes(&self) -> Vec<u8> {
-        self.iter().flat_map(|pid| pid.0.to_le_bytes()).collect()
+/// The SharedContext contains fixed known parameters accross the entire
+/// protocol. It does not however contain the entire protocol context.
+pub struct SharedContext {
+    participants: Vec<ParticipantIdentifier>,
+    generator: CurvePoint,
+    order: BigNumber,
+}
+impl ProofContext for SharedContext {
+    fn as_bytes(&self) -> Result<Vec<u8>> {
+        match bincode::serialize(&self.generator) {
+            Ok(ser_gen) => Ok([
+                self.participants
+                    .iter()
+                    .flat_map(|pid| pid.0.to_le_bytes())
+                    .collect(),
+                ser_gen,
+                self.order.to_bytes(),
+            ]
+            .concat()),
+            Err(_e) => Err(InternalError::InternalInvariantFailed),
+        }
```

Comment:
suggestion: This is fine as-is, but you could also use the `map_err` function to simplify the match here, like
`bincode::serialize(&self.generator).map_err(|_| InternalError::InternalInvariantFailed)?`